File: code/client/web/src/Domain/wrappers/identifier/IdentifierValidationError.ts
export class IdentifierValidationError {
    private readonly defaultMessage: string;

    constructor(defaultMessage: string = 'Identifier is invalid') {
        this.defaultMessage = defaultMessage;
    }

    static NegativeValue = new (class extends IdentifierValidationError {
        constructor() {
            super('Identifier cannot be negative');
        }
    })();

    toErrorMessage(): string {
        return this.defaultMessage;
    }
}

File: code/server/jvm/im-domain/src/main/kotlin/im/domain/wrappers/identifier/IdentifierValidationError.kt
package im.domain.wrappers.identifier

sealed class IdentifierValidationError(
    private val defaultMessage: String = "Identifier is invalid",
) {
    data object NegativeValue : IdentifierValidationError("Identifier value must be positive")

    fun toErrorMessage(): String = defaultMessage
}

fun List<IdentifierValidationError>.toErrorMessage(): String = joinToString("\n") { it.toErrorMessage() }
